plugins/CustomDimensions/RecordBuilders/CustomDimension.php

```diff
@@ -0,0 +1,280 @@
+use Piwik\Tracker;
+
+class CustomDimension extends RecordBuilder
+{
+    /**
+     * @var array
+     */
+    private $dimensionInfo;
+
+    /**
+     * @var int
+     */
+    private $rankingQueryLimit;
+
+    public function __construct(array $dimensionInfo)
+    {
+        parent::__construct();
+
+        $this->dimensionInfo = $dimensionInfo;
+
+        $this->maxRowsInTable = Config::getInstance()->General['datatable_archiving_maximum_rows_custom_dimensions'];
+        $this->maxRowsInSubtable = Config::getInstance()->General['datatable_archiving_maximum_rows_subtable_custom_dimensions'];
+        $this->columnToSortByBeforeTruncation = Metrics::INDEX_NB_VISITS;
+        $this->rankingQueryLimit = $this->getRankingQueryLimit();
+    }
+
+    public function getRecordMetadata(ArchiveProcessor $archiveProcessor): array
+    {
+        $recordName = Archiver::buildRecordNameForCustomDimensionId($this->dimensionInfo['idcustomdimension']);
+        return [
+            Record::make(Record::TYPE_BLOB, $recordName),
+        ];
+    }
+
+    protected function aggregate(ArchiveProcessor $archiveProcessor): array
+    {
+        $dimension = $this->dimensionInfo;
+        if (!$dimension['active']) { // sanity check
+            return [];
+        }
+
+        $logAggregator = $archiveProcessor->getLogAggregator();
+
+        $report = new DataTable();
+        $recordName = Archiver::buildRecordNameForCustomDimensionId($dimension['idcustomdimension']);
+
+        $valueField = LogTable::buildCustomDimensionColumnName($dimension);
+        $dimensions = array($valueField);
+
+        if ($dimension['scope'] === CustomDimensions::SCOPE_VISIT) {
+            $this->aggregateFromVisits($report, $logAggregator, $valueField, $dimensions, " log_visit.$valueField is not null");
+            $this->aggregateFromConversions($report, $logAggregator, $valueField, $dimensions, " log_conversion.$valueField is not null");
+        } elseif ($dimension['scope'] === CustomDimensions::SCOPE_ACTION) {
+            $this->aggregateFromActions($report, $logAggregator, $valueField);
+        }
+
+        $report->filter(DataTable\Filter\EnrichRecordWithGoalMetricSums::class);
+
+        return [
+            $recordName => $report,
+        ];
+    }
+
+    private function aggregateFromVisits(DataTable $report, LogAggregator $logAggregator, string $valueField,
+                                         array $dimensions, string $where)
+    {
+        if ($this->rankingQueryLimit > 0) {
+            $rankingQuery = new RankingQuery($this->rankingQueryLimit);
+            $rankingQuery->addLabelColumn($dimensions[0]);
+
+            $query = $logAggregator->queryVisitsByDimension($dimensions, $where, [], false, $rankingQuery, false, -1,
+                $rankingQueryGenerate = true);
+        } else {
+            $query = $logAggregator->queryVisitsByDimension($dimensions, $where);
+        }
+
+        $defaultColumns = DataArray::makeEmptyRow();
```

Comment:
Aren't we tryng to get rid of the DataArray stuff? Wondering if there's already another method that could be used instead.

Reply:
Currently there isn't. There might be if the simpler log aggregation poc is accepted. That won't be discussed for a while though. Replaced with just setting the column values to 0 directly.